src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java

```diff
@@ -605,14 +614,75 @@ public void testCheckUpstreamBuildNumber() throws Exception {
+        assertEquals(FormValidation.Kind.OK, d.doCheckUpstreamBuildNumber(null, "project2", build1.getDisplayName()).kind);
+
+        assertEquals(FormValidation.Kind.ERROR, d.doCheckUpstreamBuildNumber(null, "project2", "9999").kind);
+        assertEquals(FormValidation.Kind.ERROR, d.doCheckUpstreamBuildNumber(null, "project2", "NosuchBuild").kind);
+
         // permission check
         Authentication a = Jenkins.getAuthentication();
         try {
             SecurityContextHolder.getContext().setAuthentication(User.get("devel").impersonate());
             assertEquals(FormValidation.Kind.OK, d.doCheckUpstreamBuildNumber(project1, "project3", "nosuchbuild").kind);  // limitation
+            assertEquals(FormValidation.Kind.OK, d.doCheckUpstreamBuildNumber(null, "project3", "nosuchbuild").kind);
         } finally {
             SecurityContextHolder.getContext().setAuthentication(a);
         }
     }
+
+    @Test
+    public void testAutoCompleteUpstreamProjectName() throws Exception {
+        DownstreamBuildSelector.DescriptorImpl d = (DownstreamBuildSelector.DescriptorImpl) j.jenkins.getDescriptorOrDie(DownstreamBuildSelector.class);
+
+        ProjectMatrixAuthorizationStrategy pmas = new ProjectMatrixAuthorizationStrategy();
+        pmas.add(Jenkins.READ, "devel");
+
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(pmas);
+
+        // project1
+        // project2  cannot read from devel
+        FreeStyleProject project1 = j.createFreeStyleProject("project1");
+        {
+            Map<Permission, Set<String>> map = new HashMap<Permission, Set<String>>();
+            map.put(Item.READ, Sets.newHashSet("devel"));
+            project1.addProperty(new AuthorizationMatrixProperty(map));
+        }
+
+        FreeStyleProject project2 = j.createFreeStyleProject("project2");
+
+        //Check Empty strings
+        testAutoCompleteUpstreamProjectName(new String [] {project1.getName(), project2.getName()}, "", project1, d);
+        //Check simple matching string
+        testAutoCompleteUpstreamProjectName(new String [] {project1.getName(), project2.getName()}, "proj", project1, d);
+        //Check non matching string
+        testAutoCompleteUpstreamProjectName(new String [] {}, "FOO", project1, d);
+        //Check matching string
+        testAutoCompleteUpstreamProjectName(new String [] {project1.getName()}, "project1", project2, d);
+    }
+
+    private void testAutoCompleteUpstreamProjectName(
+            String [] expectedValues,
+            String value,
+            AbstractProject project,
+            DownstreamBuildSelector.DescriptorImpl d) {
+
+        assertArrayEquals(expectedValues, d.doAutoCompleteUpstreamProjectName(value, project).getValues().toArray());
+        //JENKINS-32526
+        assertArrayEquals(expectedValues, d.doAutoCompleteUpstreamProjectName(value, null).getValues().toArray());

```

Comment:
It's better to sort the list or wrap it with `Set`.
`AutoCompleteCandidates` doesn't look ensure the order of values (strange behavior!)